### ISSUE TITLE
Push Notifications getting too many "504 Gateway Time-out" errors

### ISSUE THREAD

**Issue by BrodaNoel** [expo/expo#18650]
### Summary

Since February, I'm getting A LOT of errors responses while sending push notifications using Expo Notifications Tokens in the server.

I'm getting exactly this error:

```
Error: Expo responded with an error with status code 504: <html>
<head><title>504 Gateway Time-out</title></head>
<body>
<center><h1>504 Gateway Time-out</h1></center>
<hr><center>nginx</center>
</body>
</html>

    at Expo.<anonymous> (/workspace/node_modules/expo-server-sdk/build/ExpoClient.js:234:30)
    at Generator.next (<anonymous>)
    at /workspace/node_modules/expo-server-sdk/build/ExpoClient.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/workspace/node_modules/expo-server-sdk/build/ExpoClient.js:4:12)
    at Expo.getTextResponseErrorAsync (/workspace/node_modules/expo-server-sdk/build/ExpoClient.js:233:16)
    at Expo.<anonymous> (/workspace/node_modules/expo-server-sdk/build/ExpoClient.js:222:35)
    at Generator.next (<anonymous>)
    at fulfilled (/workspace/node_modules/expo-server-sdk/build/ExpoClient.js:5:58)
    at runMicrotasks (<anonymous>) 
```

And some times also this other one:

```
Error: Expo responded with an error with status code 504: <html>
<head><title>504 Gateway Time-out</title></head>
<body>
<center><h1>504 Gateway Time-out</h1></center>
<hr><center>nginx</center>
</body>
</html>

    at Expo.<anonymous> (/workspace/node_modules/expo-server-sdk/build/ExpoClient.js:234:30)
    at Generator.next (<anonymous>)
    at /workspace/node_modules/expo-server-sdk/build/ExpoClient.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/workspace/node_modules/expo-server-sdk/build/ExpoClient.js:4:12)
    at Expo.getTextResponseErrorAsync (/workspace/node_modules/expo-server-sdk/build/ExpoClient.js:233:16)
    at Expo.<anonymous> (/workspace/node_modules/expo-server-sdk/build/ExpoClient.js:222:35)
    at Generator.next (<anonymous>)
    at fulfilled (/workspace/node_modules/expo-server-sdk/build/ExpoClient.js:5:58)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) 
```

<img width="1437" alt="Screen Shot 2022-08-15 at 19 52 13" src="https://user-images.githubusercontent.com/1837969/184734117-8b830f5f-8b98-4228-8752-b8da17680ea6.png">

<img width="1440" alt="Screen Shot 2022-08-15 at 19 52 22" src="https://user-images.githubusercontent.com/1837969/184734130-57e7b667-19ef-4204-94f8-63fb4367ad12.png">

There is a huge blocking issue with this: If the response is an Error 504, I don't get the Tickets, so I can't know if the push notifications where actually sent or not.

I'm sending notifications using Chunks, and seems like when I get that error, SOME of the notifications in the bunch get send, and some others not, but it's impossible to know it. So, what I have to do is to re-send the bunch again, resulting in duplicated push notifications (that as you may imagine, users are complaining a lot)

You may thing that a way to fix this problem is:
- Send the notifications slowly, one by one instead of sending them in parallel,
- Wait some exponential time after every bunch sent
- Not sending notifications in "busy" times (as it hours XX:00)

I am already doing all that things but I still have the problem.

During the last 30 days I even implemented a change in the app to totally avoid sending more than 1 bunch of notifications every 3 minutes.

So, every 3 minutes I check if there are notifications to send, and I only select 20 or 50 of them, I create a chunk, and I send it. Then I repeat the process 3 minutes after... And I still have that problem. Only 1 call every 3 minutes with 50 tokens as much! That's really really weird.

I don't know what else I can do to reduce this problem.

If you implement an endpoint where I can check if the server is busy, I am totally willing to use it, in order to not add more notifications and wait some minutes.

I am even willing to pay for a premium server where I can ensure that it's never busy and I know the notifications are not going to be duplicated.

### What platform(s) does this occur on?

Not Applicable

### Environment

Not applicable

### Minimal reproducible example

Not applicable. I guess, but this is my code.

In my app I have some "scheduled" notifications. something like: "Send this notification in 30 hours from now", that's why I query notifications to the DB.

```js
import { Expo } from 'expo-server-sdk'; // expo-server-sdk@3.6.0 right noe

const processScheduledNotifications = async (_req, res) => {
  try {
    const expo = new Expo();

    const NOW = Date.now();
    const tinutesRightNow = new Date().getMinutes();

    const notificationsInDB = await actionsPushNotifications.getBy({
      where: [
        { _state: 1 }, // notifications pendings to be sent.
        ['data.sendAt', '<', NOW + 1] // get notifications that needs to be send now.
      ],
      limit: tinutesRightNow > 20 ? 50 : 20, // between xx:00 and xx:20 I send only 20 notifications, and 50 after that.
    });

    if (notificationsInDB.length === 0) {
      return res.send({ status: 'ok' });
    }

    const messagesToExpo = [];

    // Just building the regular notification object
    notificationsInDB.forEach(noti => {
      const message = {
        to: noti.data.token,
        title: noti.data.title,
      };

      if (noti.data.body) {
        message.body = noti.data.body;
      }

      if (noti.data.data) {
        message.data = noti.data.data;
      }

      messagesToExpo.push(message);
    });

    // Creating the chunk, which is actually only gonig to be 1 chunk with 20 or 50 as max.
    const chunks = expo.chunkPushNotifications(messagesToExpo);

    // Here I'm going to save the Tickets.
    const expoResponses = [];

    // This for each doesn't makes sense. It's only 1 chunk... but... will refactor later.
    for (let chunk of chunks) {
      try {
        // THIS IS FAILING!
        const ticketChunk = await expo.sendPushNotificationsAsync(chunk);
        expoResponses.push(...ticketChunk);
      } catch (error) {
        // If fails, just create some fake tickets with data letting me know that they failed, and add that data to the DB object.
        expoResponses.push(
          ...new Array(chunk.length).fill({
            errorWhileSending: true,
            error: error.message,
          })
        );
      }
    }

    // Now, I will add all those tickets to each notification that I wanted to send
    const proms = [];
    expoResponses.forEach((expoResponse, i) => {
      // The `expoReponse` is the ticket from Expo (if doesn't fail), or the faked object created in the catch.
      const newNotiData = {
        expoResponse,
      };

      // If there as no error, change the `_state` to 2.
      if (!expoResponse.errorWhileSending) {
        newNotiData._state = 2;
      }

      // Just add the Ticket info in the DB
      proms.push(
        actionsPushNotifications.editById(notificationsInDB[i].id, newNotiData)
      );
    });

    await Promise.all(proms);

    return res.send({ status: 'ok' });
  } catch (error) {
    return res
      .status(500)
      .send({ status: 'error', message: 'Error sending notifications: ' + error });
  }
};
```

**Comment by BrodaNoel** [comment]
Just for the records, I got this again (8/22/22, 2:48 PM Argentina Time), in a small chunk (less than 30 tokens).

```
Error: Expo responded with an error with status code 504: <html> <head><title>504 Gateway Time-out</title></head> <body> <center><h1>504 Gateway Time-out</h1></center> <hr><center>nginx</center> </body> </html>

at .Expo.<anonymous> ( /workspace/node_modules/expo-server-sdk/build/ExpoClient.js:234 )
at .Generator.next
at undefined. ( /workspace/node_modules/expo-server-sdk/build/ExpoClient.js:8 )
at .Promise
at .__awaiter ( /workspace/node_modules/expo-server-sdk/build/ExpoClient.js:4 )
at .Expo.getTextResponseErrorAsync ( /workspace/node_modules/expo-server-sdk/build/ExpoClient.js:233 )
at .Expo.<anonymous> ( /workspace/node_modules/expo-server-sdk/build/ExpoClient.js:222 )
at .Generator.next
at .fulfilled ( /workspace/node_modules/expo-server-sdk/build/ExpoClient.js:5 )
at .processTicksAndRejections ( [node:internal/process/task_queues:96](https://console.cloud.google.com/debug?referrer=fromlog&file=node:internal%2Fprocess%2Ftask_queues&line=96&project=mari-a5cc7) )
```

**Comment by Jojocaster** [comment]
I also get a bunch of 504s. However, they seem to get sent anyway. I've had many users reporting duplicate notifications, but there is no other way for us to know whether they've actually been sent or not, given that we are not getting the receipts back in case of a 504. 
We had to disable retries for now, but it would be good to find some kind of solution. 

I do appreciate that expo-notifications is a free service, and that no free service can ever be perfect, but it would be great if we had a way to check whether these were actually sent, queued, or if we did actually get a 504 🙂

**Comment by Firzenizer** [comment]
We have also been experiencing a lot of 504s in the past few days. It seems there is nothing we can do.

**Comment by BrodaNoel** [comment]
Since 1 week ago, I'm seeing that some of the Expo Responses I get when I call `sendPushNotificationsAsync`, doesn't contain the `id`, which is the `receiptId`.

Never happened in 3 years of using it, and now it already happened 2 times in the last 7 days.

I hope that's something related to some fix Expo Team is trying to push :D

**Comment by jackkinsella** [comment]
I started experienced a bunch of these in the last week

**Comment by BrodaNoel** [comment]
More reports. Argentina time:

<img width="669" alt="Screen Shot 2022-09-21 at 12 57 36" src="https://user-images.githubusercontent.com/1837969/191553209-6435b8bf-47d6-4daf-9a8e-8b4644a39456.png">

**Comment by pganster** [comment]
We continuously have the same problem with the gateway timeouts and it seems to get worse over the last weeks. I don't know if Expo even has this on their radar, as the [status page](https://status.expo.dev/) doesn't seem to report any problems with their push notifications broker.

In the last week our **hourly job** that sends push notifications failed 16 times with a 504 gateway timeout and just yesterday it has failed **seven times**. 

All in all, the push notification service by Expo is a great concept, but it is now time that Expo offers a paid service with an uptime guarantee.

**Comment by atwiltshire** [comment]
I had the same issue back in January and raised a question on the [expo forums](https://forums.expo.dev/t/error-thrown-while-updating-the-device-push-token-with-the-server/61323), after a few days the issue seemed to be resolved. However, over the weekend it has returned, I have asked again on the [expo forum](https://forums.expo.dev/t/error-thrown-while-calling-notifications-getexpopushtokenasync/66676).

**Comment by Noitidart** [comment]
Same situation here. In the last 24h though the errors are 600% higher for me.

**Comment by BrodaNoel** [comment]
My recommendation: If the expo call to send a notification retrieves an error, consider it as it was send.

During the weekend the error happened 8 times in a row. 18% of my users disabled the notifications because I was flooding them.

Try to avoid making the same mistake I did. I already disabled the "retry" behavior.

<img width="102" alt="Screen Shot 2022-10-03 at 14 32 12" src="https://user-images.githubusercontent.com/1837969/193641630-a400038f-dcc0-43f1-abcf-4615fa242987.png">

**Comment by briefjudofox** [comment]
In recent weeks, we've also experienced what others have described in this thread (timeouts for clients getting a token and timeouts on the endpoints to send notifications and fetch tickets).  FWIW, we basically arrived to the same conclusion as @BrodaNoel.  

- Remove retry logic for backend timeouts (we've kept retry logic for other types of errors like connection closed, connection refused, 502, & 503 errors)
- Remove our own network request timeout for calls to the expo push service (let the expo service time the request out)

In the event of a timeout without a response with receipt(s) it's tricky to confirm that a push was delivered.  However, we had a couple of instances where members of our team received duplicate push notifications and we were able to tie them to timeouts on our logs.  We were also able to do that same for some users who reported receiving duplicate notifications.

**Comment by Noitidart** [comment]
> My recommendation: If the expo call to send a notification retrieves an error, consider it as it was send.
> 
> During the weekend the error happened 8 times in a row. 18% of my users disabled the notifications because I was flooding them.
> 
> Try to avoid making the same mistake I did. I already disabled the "retry" behavior.
> 
> <img alt="Screen Shot 2022-10-03 at 14 32 12" width="102" src="https://user-images.githubusercontent.com/1837969/193641630-a400038f-dcc0-43f1-abcf-4615fa242987.png">

Same situation here. I reduced my retries to just to one though. So worst case users get 2 notifications.

**Comment by BrodaNoel** [comment]
I just discovered that since 2 days ago I'm having a HUGE amount of "Device not registered" answers from the ticketId. Which ended up in my code removing that Token from the DB, but I just realized that those may have been false-positives.
Why? Because I just sent a bunch of 90 push notifications, and I got the ticket information from those, and 100% of them had the "Device not registered" status. Quite weird, considering that usually it's less than 5% (users who removed the app).

Are some of you having the same problem?

From the 100% of the users I have in my DB, I always had 22% of them with valid/active ExpoTokens, and since 3 days ago it moved down to 17.8%... Feels weird, because I'm not seeing users uninstaling the app.

Maybe another issue with the push notification service?

**Comment by Noitidart** [comment]
I don't think I got an unexpected amount. I think my users are really uninstalling the app.

I do have issues with some users complaining they don't get notifications. I'm not sure what can cause this.

**Comment by BrodaNoel** [comment]
Well... This is causing it... 6 hours ago I got the same.

<img width="837" alt="Screen Shot 2022-10-07 at 22 12 27" src="https://user-images.githubusercontent.com/1837969/194680221-5cbfefe6-d404-4f13-b5d9-e2d33e942110.png">

**Comment by Noitidart** [comment]
This has picked up again recently, this is the error I'm seeing often:

```
Error: Expo responded with an error with status code 503: upstream connect error or disconnect/reset before headers. reset reason: connection failure, transport failure reason: delayed connect error: 111
    at Expo.<anonymous> (/opt/apps/themasjidapp/node_modules/expo-server-sdk/build/ExpoClient.js:189:28)
    at Generator.next (<anonymous>)
    at /opt/apps/themasjidapp/node_modules/expo-server-sdk/build/ExpoClient.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/opt/apps/themasjidapp/node_modules/expo-server-sdk/build/ExpoClient.js:3:12)
    at Expo._getTextResponseErrorAsync (/opt/apps/themasjidapp/node_modules/expo-server-sdk/build/ExpoClient.js:188:16)
    at Expo.<anonymous> (/opt/apps/themasjidapp/node_modules/expo-server-sdk/build/ExpoClient.js:177:35)
    at Generator.next (<anonymous>)
    at fulfilled (/opt/apps/themasjidapp/node_modules/expo-server-sdk/build/ExpoClient.js:4:58)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by BrodaNoel** [comment]
I'm also getting a couple of:

```
Expo responded with an error with status code 503: upstream connect error or disconnect/reset before headers. reset reason: connection failure, transport failure reason: delayed connect error: 111
```

**Comment by BrodaNoel** [comment]
Today I sent 9.000 push notifications and all of them went good without any error.
Are some of you still having this problem? Maybe it got fixed.

**Comment by Noitidart** [comment]
@BrodaNoel I havent seen any errors myself for last couple days. Praying its fixed haha.

**Comment by Noitidart** [comment]
Got a couple 503's but its much less then the many we used to get in past.

```
Expo responded with an error with status code 503: upstream connect error or disconnect/reset before headers. reset reason: connection termination
```

**Comment by Noitidart** [comment]
Got a buuuunch of the 503's just now:

> Expo responded with an error with status code 503: upstream connect error or disconnect/reset before headers. reset reason: connection termination

**Comment by Noitidart** [comment]
These are some errors I have seen thrown by `sendPushNotificationAsync`. Does anyone know have any recommendations how to handle each one? Currently I only retry on the `TOO_MANY_REQUESTS` error, and for the others I don't retry as I think it is sending, so I don't risk a duplicate notification.

## Error 1
```
  {
    "stack": "Error: Exceeded 600 notifications per second, please try again.\n    at Expo._getErrorFromResultError (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:212:21)\n    at Expo._getErrorFromResult (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:202:26)\n    at Expo.<anonymous> (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:184:25)\n    at Generator.next (<anonymous>)\n    at fulfilled (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:4:58)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)",
    "code": "TOO_MANY_REQUESTS",
    "message": "Exceeded 600 notifications per second, please try again."
  }
```

## Error 2
```
  {
    "code": "INTERNAL_SERVER_ERROR",
    "message": "An unknown error occurred.",
    "stack": "Error: An unknown error occurred.\n    at Expo._getErrorFromResultError (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:212:21)\n    at Expo._getErrorFromResult (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:202:26)\n    at Expo.<anonymous> (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:184:25)\n    at Generator.next (<anonymous>)\n    at fulfilled (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:4:58)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)"
  }
```

Error 3
```
  {
    "errorText": "upstream connect error or disconnect/reset before headers. reset reason: connection termination",
    "stack": "Error: Expo responded with an error with status code 503: upstream connect error or disconnect/reset before headers. reset reason: connection termination\n    at Expo.<anonymous> (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:189:28)\n    at Generator.next (<anonymous>)\n    at /applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:7:71\n    at new Promise (<anonymous>)\n    at __awaiter (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:3:12)\n    at Expo._getTextResponseErrorAsync (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:188:16)\n    at Expo.<anonymous> (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:177:35)\n    at Generator.next (<anonymous>)\n    at fulfilled (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:4:58)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)",
    "message": "Expo responded with an error with status code 503: upstream connect error or disconnect/reset before headers. reset reason: connection termination",
    "statusCode": 503
  }
```

Error 4
```
  {
    "errorText": "no healthy upstream",
    "stack": "Error: Expo responded with an error with status code 503: no healthy upstream\n    at Expo.<anonymous> (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:189:28)\n    at Generator.next (<anonymous>)\n    at /applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:7:71\n    at new Promise (<anonymous>)\n    at __awaiter (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:3:12)\n    at Expo._getTextResponseErrorAsync (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:188:16)\n    at Expo.<anonymous> (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:177:35)\n    at Generator.next (<anonymous>)\n    at fulfilled (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:4:58)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)",
    "message": "Expo responded with an error with status code 503: no healthy upstream",
    "statusCode": 503
  }
```

## Error 5
```
  {
    "stack": "Error: The Expo API server is temporarily unavailable due to high load. Try again in a moment.\n    at Expo._getErrorFromResultError (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:212:21)\n    at Expo._getErrorFromResult (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:202:26)\n    at Expo.<anonymous> (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:184:25)\n    at Generator.next (<anonymous>)\n    at fulfilled (/applications/my-expo-app/node_modules/expo-server-sdk/build/ExpoClient.js:4:58)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)",
    "code": "SERVICE_UNAVAILABLE",
    "message": "The Expo API server is temporarily unavailable due to high load. Try again in a moment."
  }
```
